src/wp-includes/link-template.php

```diff
@@ -417,13 +467,26 @@ function get_attachment_link( $post = null, $leavename = false ) {
 
 	$link = false;
 
-	$post   = get_post( $post );
-	$parent = ( $post->post_parent > 0 && $post->post_parent != $post->ID ) ? get_post( $post->post_parent ) : false;
-	if ( $parent && ! in_array( $parent->post_type, get_post_types(), true ) ) {
-		$parent = false;
+	$post            = get_post( $post );
+	$force_ugly_link = wp_force_ugly_post_permalink( $post );
```

Comment:
Need for sample/converting future links can be dealt with in https://core.trac.wordpress.org/ticket/52324 or https://core.trac.wordpress.org/ticket/36976